File: PetStore/PetStore/Services/ProductsService.cs
namespace PetStore.Services
{
    using System;
    using System.IO;
    using System.Linq;
    using System.Threading.Tasks;

    using PetStore.Data;
    using PetStore.Data.Models;
    using PetStore.Models;

    public class ProductsService : IProductsService
    {
        private readonly ApplicationDbContext dbContext;

        public ProductsService(ApplicationDbContext dbContext)
        {
            this.dbContext = dbContext;
        }

        public async Task<Product> CreateAsync(AddProductModel input, string imagePath)
        {
            var product = new Product
            {
                Name = input.Name,
                Price = input.Price,
                Capacity = input.Capacity,
                PetId = input.PetId,
                ProductTypeId = input.ProductTypeId,
                Description = input.Description,
                IsDeleted = false,
                CreatedOn = DateTime.UtcNow,
            };

            Directory.CreateDirectory(imagePath);
            if (input.Image != null)
            {
                var image = input.Image;
                var extension = Path.GetExtension(image.FileName).TrimStart('.');

                var productImage = new ProductImage
                {
                    Extention = extension,
                    CreatedOn = DateTime.UtcNow,
                };

                var physicalPath = $"{imagePath}{productImage.Id}.{extension}";
                using Stream fileStream = new FileStream(physicalPath, FileMode.Create);
                await image.CopyToAsync(fileStream);

                product.ProductImage = productImage;
            }

            await this.dbContext.AddAsync(product);
            await this.dbContext.SaveChangesAsync();

            return product;
        }

        public async Task DeleteAsync(string id)
        {
            var product = this.dbContext
                .Products
                .FirstOrDefault(p => p.Id == id && p.IsDeleted == false);
            if (product == null)
            {
                throw new NullReferenceException("There is no product found.");
            }

            var productImage = this.dbContext
                .Products
                .Select(p => p.ProductImage)
                .FirstOrDefault();
            if (productImage != null)
            {
                productImage.DeletedOn = DateTime.UtcNow;
                productImage.IsDeleted = true;
            }

            product.DeletedOn = DateTime.UtcNow;
            product.IsDeleted = true;

            await this.dbContext.SaveChangesAsync();
        }

        public async Task<Product> EditAsync(EditProductModel input)
        {
            var product = this.dbContext
                .Products
                .Where(p => p.Id == input.ProductId)
                .FirstOrDefault();

            product.Name = input.Name;
            product.Price = input.Price;
            product.Capacity = input.Capacity;
            product.ModifiedOn = DateTime.UtcNow;

            await this.dbContext.SaveChangesAsync();

            return product;
        }

        public ProductsListModel GetAll()
        {
            var products = this.dbContext
                .Products
                .Where(p => p.Capacity > 0 && p.IsDeleted == false)
                .OrderByDescending(p => p.CreatedOn)
                .Select(p => new ProductInListModel
                {
                    Id = p.Id,
                    Name = p.Name,
                    Price = p.Price,
                    ImagePath = p.ProductImage != null
                        ? $"/images/products/{p.ProductImageId}.{p.ProductImage.Extention}"
                        : "/images/defaults/placeholder.png",
                })
                .ToList();

            var productsList = new ProductsListModel
            {
                Products = products,
            };

            return productsList;
        }

        public SingleProductModel GetById(string id)
        {
            return this.dbContext
                .Products
                .Where(p => p.Id == id)
                .Select(p => new SingleProductModel
                {
                    Id = p.Id,
                    Description = p.Description,
                    ImagePath = p.ProductImage != null
                        ? $"/images/products/{p.ProductImageId}.{p.ProductImage.Extention}"
                        : "/images/defaults/placeholder.png",
                    Name = p.Name,
                    Price = p.Price,
                })
                .FirstOrDefault();
        }

        public EditProductModel GetByIdEditModel(string id)
        {
            return this.dbContext
                .Products
                .Where(p => p.Id == id)
                .Select(p => new EditProductModel
                {
                    ProductId = p.Id,
                    Name = p.Name,
                    Price = p.Price,
                    Capacity = p.Capacity,
                })
                .FirstOrDefault();
        }

        public ProductNameModel GetName(string id)
        {
            return this.dbContext
                .Products
                .Where(p => p.Id == id && p.IsDeleted == false)
                .Select(p => new ProductNameModel
                {
                    Name = p.Name,
                    Id = p.Id,
                })
                .FirstOrDefault();
        }
    }
}

File: PetStore/PetStore/ClientApp/src/services/productsService.js
import api from './api';

export const getAll = () => {
    return fetch(api.products)
        .then(res => res.json())
        .catch(err => console.log('Handled error:' + err));
};

export const getOne = (id) => {
    return fetch(`${api.products}/getOne?id=${id}`)
        .then(res => res.json())
        .catch(err => console.log('Handled error:' + err));
};

export const getName = (id) => {
    return fetch(`${api.products}/name?id=${id}`)
        .then(res => res.json())
        .catch(err => console.log('Handled error:' + err));
};

export const getEditModel = (id) => {
    return fetch(`${api.products}/edit?id=${id}`)
        .then(res => res.json())
        .catch(err => console.log('Handled error:' + err));
};

export const create = (productName, price, capacity, petId, productTypeId, image, description) => {
    let product = new FormData();
    product.append('name', productName);
    product.append('price', price);
    product.append('capacity', capacity);
    product.append('petId', petId);
    product.append("productTypeId", productTypeId);
    product.append("image", image);
    product.append("description", description);

    return fetch(api.products, {
        method: 'POST',
        body: product
    })
        .then(res => {
            if (res.ok) {
                return res.json();
            }

            throw new Error('Invalid input');
        });
};

export const edit = (productName, price, capacity, id) => {
    let product = new FormData();
    product.append('name', productName);
    product.append('price', price);
    product.append('capacity', capacity);
    product.append('productId', id);

    return fetch(api.products, {
        method: 'PUT',
        body: product
    })
        .then(res => {
            if (res.ok) {
                return res.json();
            }

            throw new Error('Invalid input');
        });
};

export const $delete = (productId) => {
    const product = new FormData();
    product.append("id", productId);

    return fetch(api.products, {
        method: 'DELETE',
        body: product
    })
        .catch(err => console.log('Handled error:' + err));
};
